lib/Pipes/PipelineManager.cpp

```diff
@@ -162,26 +156,187 @@ checkComponentsVersion(const revng::DirectoryPath &ExecutionDirectory,
+static llvm::Expected<revng::FilePath>
+getBackupFilePath(const revng::FilePath &Path) {
+  // Keep trying to get a non-occupied backup file path. The timestamp is
+  // precise down to microseconds, so it is really low probability this takes
+  // more than one iteration.
+  while (true) {
+    const std::string DateTime = revng::getDateTimeISOFormat();
+
+    // addExtension does not allow '.' to be in the extension, so we have to
+    // manually take it out of the ISO-formatted date-time.
+    const auto Size = DateTime.size();
+    const std::string FirstExtension = llvm::formatv("backup@{0}",
+                                                     DateTime.substr(0,
+                                                                     Size - 7));
+    const std::string SecondExtension = DateTime.substr(Size - 6);
+
+    auto BackupFilePath = Path.addExtension(FirstExtension)
+                            .addExtension(SecondExtension);
+
+    auto MaybeExists = BackupFilePath.exists();
+
+    if (not MaybeExists)
+      return MaybeExists.takeError();
+
+    if (not MaybeExists.get())
+      return BackupFilePath;
+  }
+}
+
+static llvm::Expected<revng::FilePath>
+createModelBackup(const revng::FilePath &ModelFile) {
+  llvm::Expected<revng::FilePath>
+    ModelBackupFile = getBackupFilePath(ModelFile);
+  if (not ModelBackupFile)
+    return ModelBackupFile.takeError();
+
+  if (auto Error = ModelFile.copyTo(ModelBackupFile.get()); !!Error)
+    return Error;
+
+  return ModelBackupFile;
+}
+
+static llvm::Error migrateModelFile(const revng::FilePath &ModelFile) {
+  TemporaryFile InputTemporaryFile("revng-model-migration-input", "yml");
+  TemporaryFile OutputTemporaryFile("revng-model-migration-output", "yml");
+
+  revng::FilePath
+    InputFile = revng::FilePath::fromLocalStorage(InputTemporaryFile.path());
+  if (auto Error = ModelFile.copyTo(InputFile); !!Error)
+    return Error;
+
+  {
+    int ReturnCode = ::Runner.run("revng",
+                                  { "model",
+                                    "migrate",
+                                    InputTemporaryFile.path().str(),
+                                    "--output",
+                                    OutputTemporaryFile.path().str() });
+
+    if (ReturnCode != 0)
+      return revng::createError("Error encountered while running migrations, "
+                                "process returned: %d",
+                                ReturnCode);
+  }
+
+  revng::FilePath
+    OutputFile = revng::FilePath::fromLocalStorage(OutputTemporaryFile.path());
+  if (auto Error = OutputFile.copyTo(ModelFile); !!Error)
+    return Error;
+
+  return llvm::Error::success();
+}
+
+// Conveninece function for setUpPipeline, creates a backup for the model file
+// and runs migrations on it. On success returns the backup path, on failure
+// returns the error and restores the model file from backup. If something goes
+// horribly wrong and it fails to restore, it aborts.
```

Comment:
~~Conveninece function for setUpPipeline,~~ creates a backup for the model file
and runs migrations on it.

`\return` on success returns the backup path, on failure
returns the error and restores the model file from backup. ~~If something goes
 horribly wrong and it fails to restore, it aborts.~~